[fix bug 1443566] Home page redesign A/B test

alexgibson opened the pull request:
## Description
- Adds home page redesign A/B test [en-US only, no l10n required at this stage].
- Card images are lazyloaded via `IntersectionObserver`.
- Videos are only loaded on click.
- Updates our image lazyload plugin to support `srcset`.

## Issue / Bugzilla link
https://bugzilla.mozilla.org/show_bug.cgi?id=1443566

## Testing
Demo:
- https://bedrock-demo-agibson.oregon-b.moz.works/en-US/?v=a (redesign)
- https://bedrock-demo-agibson.oregon-b.moz.works/en-US/?v=b (control)

Checklist:
- [ ] Test experiment redirect works as expected.
- [ ] Test layout responsiveness.
- [ ] Test CSS grid fallback in older browsers.
- [ ] Test videos play as expected.
- [ ] Test images load as expected.

alexgibson commented:
Successful test run: https://ci.us-west.moz.works/blue/organizations/jenkins/bedrock_multibranch_pipeline/detail/run-integration-tests/187/pipeline/

alexgibson commented on bedrock/mozorg/templates/mozorg/home/home-a.html:
```diff
@@ -0,0 +1,417 @@
+              image_url='img/home/2018/people-over-profit.png',
+              image_highres_url='img/home/2018/people-over-profit-high-res.png',
+              aspect_ratio='has-aspect-3-2',
+              link_url='https://blog.mozilla.org/blog/2018/04/05/we-stand-for-people-over-profit/?utm_source=www.mozilla.org&utm_medium=referral&utm_campaign=homepage&utm_content=v-a'
```
TODO: This blog post is still 404, check that's it's published before merging.

alexgibson commented:
OK, still waiting on some final images - but this is now close enough to start code review.

jpetto commented on bedrock/mozorg/templates/mozorg/home/home-b.html:
```diff
@@ -0,0 +1,47 @@
+    <a rel="external" href="https://research.mozilla.org/rust/?utm_source=www.mozilla.org&utm_medium=referral&utm_campaign=homepage&utm_content=v-b"><svg class="icon"><use xlink:href="#icon-rust"></use></svg>{{ _('Rust') }}</a>
+  </li>
+</ul>
+{% endblock %}
```
Is this content different than what's currently served to `en-US` visitors in `home.html`? Maybe I just can't see it...

alexgibson replied:
The only difference are the `utm` params in the links, which is something required for GA.

alexgibson commented on bedrock/mozorg/templates/mozorg/home/home-a.html:
```diff
@@ -0,0 +1,417 @@
+        <source src="{{ video_url }}.mp4" type="video/mp4">
+      </video>
+      <figcaption>
+        <p>{{ desc }} <a href="{{ link_url }}">{{ _('Read more') }}</a></p>
```
Good catch 👍

alexgibson commented on bedrock/base/templates/macros.html:
```diff
@@ -35,11 +35,11 @@
 {%- endmacro %}
 
-{% macro lazy_image(image_url, placeholder_url) -%}
+{% macro lazy_image(image_url, image_highres_url, placeholder_url, image_class) -%}
```
Will tak a look and see how easy this is in Jinja 👍

alexgibson commented on bedrock/mozorg/templates/mozorg/home/home-a.html:
```diff
@@ -0,0 +1,417 @@
+
+{% macro card_block(class, tag_label, tag_class, title, ga_title, desc, image_url, image_highres_url, video_url, aspect_ratio, link_url, small_cta, meta) -%}
+<section class="card {% if class %}{{ class }}{% endif %} {% if video_url %}has-video{% endif %}">
+  <a class="card-block-link" href="{{ link_url }}" data-link-name="{{ ga_title }}" data-link-type="link">
```
Since we don't know if a card will always link off-site, adding this attribute would likely mean adding another parameter to the macro - which I'm trying to avoid unless really necessary. Adding `rel="external"` doesn't really have any major benefits (afaict), so I think we're OK to leave it.

jpetto replied:
Agreed 👍

alexgibson commented on bedrock/mozorg/templates/mozorg/home/home-a.html:
```diff
@@ -0,0 +1,417 @@
+            image_url='img/home/2018/pocket-2.jpg',
+            aspect_ratio='has-aspect-16-9',
+            link_url='https://www.nytimes.com/2018/04/03/style/sweater-guy-sam-barsky-landmarks.html',
+            meta=_('1 hours ago')
```
Yeah - eventually they want these cards to be dynamic. This is just placeholder for now so could well change in the future 🙄

alexgibson commented on bedrock/settings/static_media.py:
```diff
@@ -1211,10 +1217,17 @@
     'home': {
         'source_filenames': (
-            'js/libs/jquery.cycle2.min.js',
-            'js/libs/jquery.waypoints.min.js',
```
I don't think they have been - the bundle is from the old, old home page and was never removed until now 😄

alexgibson commented on bedrock/settings/static_media.py:
```diff
@@ -1211,10 +1217,17 @@
-            'js/libs/jquery.cycle2.min.js',
-            'js/libs/jquery.waypoints.min.js',
+            'js/base/mozilla-lazy-load.js',
+            'js/base/mozilla-modal.js',
```
Only the new home page added here should be loading this bundle. The previous home page wasn't loading a JS bundle.

alexgibson commented:
@jpetto thanks for the review! I took your suggestion and turned the macro into a handy Python helper (like `high_res_img`) and added some tests for it. I also pushed it to demo. I added it as a second commit for easier review.

jpetto commented on docs/mozilla-lazy-load.rst:
```diff
@@ -13,34 +13,26 @@ Mozilla Image Lazy Loader
 
-Required Parameters
-~~~~~~~~~~~~~~~~~~~
+    {{ lazy_img((image_url='path/to/image.png', placeholder_url='path/to/placeholder.png', include_highres_image=True, optional_attributes={'class': 'some-class-name', 'alt': ''}) }}
```
Looks like an extra `(` snuck in there.

jpetto commented on bedrock/base/templates/macros.html:
```diff
@@ -35,15 +35,6 @@
   {% endfor %}
 {%- endmacro %}
 
-{% macro lazy_image(image_url, image_highres_url, placeholder_url, image_class) -%}
```
Looks like this is still imported in `firefox/features/index.html`. It doesn't appear to be actually used in that file, and also doesn't cause an error as-is, but still probably best to remove the import. 😉

jpetto commented on bedrock/mozorg/templatetags/misc.py:
```diff
@@ -226,6 +226,36 @@ def high_res_img(ctx, url, optional_attributes=None):
+    else:
+        class_name = 'lazy-image'
+        attrs = ''
+
```
Maybe add a check here for an `alt` optional attribute, and, if not present, default it to an empty string? Could save a fair amount of repetition in the template calls.

alexgibson commented:
@jpetto pushed some review fixes

stephaniehobson commented:
🎉 Wow! Excited to see this test start.

alexgibson commented:
@jpetto added a commit with some final design tweaks / images. This is now good to go once reviewed 🎉
